### PR TITLE
Multisample framebuffer

### PR DESCRIPTION
**Pull request by mgsx-dev**
This is an implementation proposal for framebuffer multisample feature.

When using multisampling with framebuffers, you need 2 OpenGL FBOs : 
* a mutisample FBO used to capture the scene. This one can't be rendered directly or used as regular textures.
* a non multisample FBO where resolved samples will be copied to textures and then usable as regular textures.

There was several ways to implementing it. There was an old PR where GLFrameBuffer was handling 2 OpenGL framebuffers. This overcomplexify GLFrameBuffer code with a lot of code duplication. It also makes it more complex to configure and introduce confusion : (eg. which one is bound when calling bind method). So i decided to not take that route.

This PR use an approach closer to the OpenGL API. As you can see in the test, user code have to create 2 FrameBuffers and call the transfer method to copy multisamples into the other FBO. Another benefit is that in some case (hardware specific) it's not possible to have a texture for the stencil buffer or the packed depth and stencil buffer. With this approach, you can render a scene using both depth and stencil buffers, and later transfer only depth pixels to a texture. It also allows internal format conversion (eg. rendering colors into a multisample float buffer and transfer to an RGBA8 texture).

Note that, it could be possible to have texture attachments with multisample enabled. That allows to get hand on samples in order to compose with other multisampled textures and handle antialiasing manually. These textures can't be rendered directly neither. This feature (a very advanced case imho) would requires more changes and i thought it didn't worth it at the moment. An exception is thrown when using multisample FBO and texture attachement together.

Transfer code is a bit complex because of OpenGL API apparently designed for non MRT context. `glBlitFramebuffer` is the command that perform the copy, it allows to copy depth, stencil and one color buffer in a single call. But if you have multiple color attachement, it's necessary to call it again for other color buffers.

There are surely some stuff to improve, i'd like to have your opinion on the API proposal.

### CODE REVIEW

**Comment by mgsx-dev** [comment]
> When running the `GL31FrameBufferMultisampleMRTTest` on Windows Desktop I am just seeing a black screen. Is this normal?

No, it should looks like this : 
Screenshot 2023-07-21 04:41:21

Maybe on Windows it requires `glEnable(GL_MULTISAMPLE);` ? from [this tutorial](https://learnopengl.com/Advanced-OpenGL/Anti-Aliasing) : 

> On most OpenGL drivers, multisampling is enabled by default so this call is then a bit redundant, but it's usually a good idea to enable it anyways.

Could you try to do that ? it should be done in desktop code, because it's not part of GLES : `glEnable(GL13.GL_MULTISAMPLE);`

**Comment by JamesTKhan** [comment]
> 

No luck, I did try that yesterday too. Tried again just to be sure, tried both GL13 ENUM as well as just plugging its integer value in directly.
`Gdx.gl.glEnable(GL13.GL_MULTISAMPLE);`
`Gdx.gl.glEnable(0x809D);` 

I also tried putting GLProfiler into the test, but not getting any error messages out.

On Windows 11 with Nvidia 3080

image

**Comment by JamesTKhan** [comment]
Narrowed the problem down. If I remove `Gdx.gl.glEnable(GL20.GL_DEPTH_TEST);` in render method of test, it works.

**Comment by mgsx-dev** [comment]
> Narrowed the problem down. If I remove `Gdx.gl.glEnable(GL20.GL_DEPTH_TEST);` in render method of test, it works. Or I can keep the depth test, but I must add true to the first ScreenUtils clear method to clear depth as the first one does not pass true.
> 
> `ScreenUtils.clear(Color.CLEAR, true);`

Good catch! it's fixed. I remember now that default frame buffer clearing have difference between windows and linux. Thanks.

**Comment by JamesTKhan** [comment]
> > Narrowed the problem down. If I remove `Gdx.gl.glEnable(GL20.GL_DEPTH_TEST);` in render method of test, it works. Or I can keep the depth test, but I must add true to the first ScreenUtils clear method to clear depth as the first one does not pass true.
> > `ScreenUtils.clear(Color.CLEAR, true);`
> 
> Good catch! it's fixed. I remember now that default frame buffer clearing have difference between windows and linux. Thanks.

Just checking back in. Your latest changes are working for me on the test. I submitted my other two concerns/observations as part of the GH review process.

**Comment by xzaleksey** [comment]
Any updates?)

**Comment by antzGames** [comment]
Hey folks.  I check in every week to see if they are any updates on this PR.  I am sure many are also waiting on this.  

I know its summer and we are all busy.   A friendly reminder that the community appreciates the work you do.  Cheers!

**Comment by antzGames** [comment]
What can the community do to help the libGDX contributors reduce the cycle time for PRs to get merged?  This is an honest and sincere question.

**Comment by tommyettinger** [comment]
> What can the community do to help the libGDX contributors reduce the cycle time for PRs to get merged? This is an honest and sincere question.

I don't know enough about this domain to be able to adequately review this PR. I think we're kind-of short on people who do have that domain knowledge, much like we are with people who understand Tiled PRs.

File: gdx/src/com/badlogic/gdx/graphics/glutils/GLFrameBuffer.java [hunk 1]
```diff
@@ -350,6 +401,71 @@ public void end (int x, int y, int width, int height) {
...
+			attachmentIndex++;
+		}
+		// case of depth and/or stencil only
+		if (copyBits != GL20.GL_COLOR_BUFFER_BIT) {
+			Gdx.gl30.glBlitFramebuffer(0, 0, getWidth(), getHeight(), 0, 0, destination.getWidth(), destination.getHeight(),
+				copyBits, GL20.GL_NEAREST);
+		}
+
+		// restore draw buffers for destination
+		Gdx.gl30.glDrawBuffers(destination.defaultDrawBuffers.limit(), destination.defaultDrawBuffers);
```
Comment (JamesTKhan) [inline gdx/src/com/badlogic/gdx/graphics/glutils/GLFrameBuffer.java:463]: When testing out the FBO transfer method, I was getting a NPE crash if transferring to a regular FBO that is not MRT.  `defaultDrawBuffers` is only available on FBO that has MRT or samples enabled. 

Locally I added a null check like so, and the transfer now works for non-MRT destination FBOS

```
if (destination.defaultDrawBuffers != null) {
	// restore draw buffers for destination
	Gdx.gl30.glDrawBuffers(destination.defaultDrawBuffers.limit(), destination.defaultDrawBuffers);
}
```

**Review by JamesTKhan — CHANGES_REQUESTED** [review]

**Comment by obigu** [comment]
@mgsx-de It'd be great if you could have a look at @JamesTKhan review comment https://github.com/libgdx/libgdx/pull/7185#discussion_r1345826056 when you have the chance, considering the interest we should aim at getting it merged in 1.12.2.

File: gdx/src/com/badlogic/gdx/graphics/glutils/GLFrameBuffer.java [hunk 2]
```diff
@@ -350,6 +401,71 @@ public void end (int x, int y, int width, int height) {
...
+		for (FrameBufferTextureAttachmentSpec attachment : destination.bufferBuilder.textureAttachmentSpecs) {
+			if (attachment.isColorTexture()) {
+				Gdx.gl30.glReadBuffer(GL30.GL_COLOR_ATTACHMENT0 + colorBufferIndex);
+
+				singleInt.clear();
+				singleInt.put(GL30.GL_COLOR_ATTACHMENT0 + attachmentIndex);
+				singleInt.flip();
+				Gdx.gl30.glDrawBuffers(1, singleInt);
+
+				Gdx.gl30.glBlitFramebuffer(0, 0, getWidth(), getHeight(), 0, 0, destination.getWidth(), destination.getHeight(),
```
[the remark this answers is not in the document]
Comment (mgsx-dev) [reply gdx/src/com/badlogic/gdx/graphics/glutils/GLFrameBuffer.java:453]: indeed, i added a size check and throw an error

File: gdx/src/com/badlogic/gdx/graphics/glutils/GLFrameBuffer.java [hunk 3]
```diff
@@ -350,6 +401,71 @@ public void end (int x, int y, int width, int height) {
...
+	 * method that automatically choose which of stencil, depth, and colors buffers attachment to be copied.
+	 * @param destination the destination of the copy. */
+	public void transfer (GLFrameBuffer<T> destination) {
+
+		int copyBits = 0;
+		for (FrameBufferTextureAttachmentSpec attachment : destination.bufferBuilder.textureAttachmentSpecs) {
+			if (attachment.isDepth && bufferBuilder.hasDepthRenderBuffer || bufferBuilder.hasPackedStencilDepthRenderBuffer) {
+				copyBits |= GL20.GL_DEPTH_BUFFER_BIT;
+			} else if (attachment.isStencil && bufferBuilder.hasStencilRenderBuffer
+				|| bufferBuilder.hasPackedStencilDepthRenderBuffer) {
```
[the remark this answers is not in the document]
Comment (mgsx-dev) [reply gdx/src/com/badlogic/gdx/graphics/glutils/GLFrameBuffer.java]: indeed, the conditions were wrong, it's fixed now. Nice catch!

File: gdx/src/com/badlogic/gdx/graphics/glutils/GLFrameBuffer.java (hunk 1 above)
Comment (mgsx-dev) [reply gdx/src/com/badlogic/gdx/graphics/glutils/GLFrameBuffer.java]: indeed, i updated GL31FrameBufferMultisampleTest to use the real implementation and confirmed the issue. It's fixed now.

**Comment by mgsx-dev** [comment]
@JamesTKhan i think i addressed all your points (thank you for the review), it seems that you need to review again unblock the merge.

@obigu I think it's ready to be merged.

Thank you all for your patience

**Review by JamesTKhan — DISMISSED** [review]
Thank you!

**Pull request merged**